lib/hamt.ml

```diff
@@ -716,3 +698,14 @@ struct
     let ( <-- ) hamt (k, v) = add k v hamt
   end
 end
+
+module Make' = Make (StdConfig)
+
+module With_polymorphic (M : sig type t end) = struct
+  include M
+  let equal = Stdlib.(=)
+  let hash = Hashtbl.hash
+end
+
+module String = Make' (With_polymorphic (struct type t = string end))
+module Int = Make' (With_polymorphic (struct type t = int end))
```

Comment:
Couldn't we use a specialised equality for these ones? Maybe not in the scope of this PR though.

Reply:
I think this is in general a good idea. Unfortunately there is no monomorphic `equal` for `string` before OCaml 4.03 nor `int`s before OCaml 4.08. There are no monomorphic `hash` functions either. An alternative could be to use `Base` which has both production-tested `equal` and `hash` functions for all built-in OCaml data types. Or implementing them on our own, but I'd rather not duplicate effort given equality performance and hashing quality is pretty important.

This particular code in question I basically just copied from @rgrinberg's branch. Fortunately, the fact that these are polymorphic can be completely shielded for the user so future improvement won't cause any breakage.

Reply:
> Fortunately, the fact that these are polymorphic can be completely shielded for the user so future improvement won't cause any breakage.

I'm fine with this. The code would probably deserve a substantive lifting anyway, and that's most probably not the worst performance pain point. Plus, let's merge your work first.